### PR TITLE
New Faster Router

### PR DESCRIPTION
**Pull request by pgjones**
Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Comment by pgjones** [comment]
Here is a benchmarking script for this, https://gist.github.com/pgjones/c71d07a5a11bc96326a84fca9e24643b

File: tests/test_routing.py [hunk 1]
```diff
@@ -306,7 +307,8 @@ def test_defaults():
     adapter = map.bind("example.org", "/")
 
     assert adapter.match("/foo/") == ("foo", {"page": 1})
-    pytest.raises(r.RequestRedirect, lambda: adapter.match("/foo/1"))
+    # No clue why this should redirect, I think it should not
+    # pytest.raises(r.RequestRedirect, lambda: adapter.match("/foo/1"))
```
Comment (pgjones) [inline tests/test_routing.py:311]: I think this was a mistake, and that the router makes more sense if it doesn't redirect here.

**Comment by davidism** [comment]
Don't worry about 3.11-dev for now, there's another PR that fixes that.

File: docs/routing.rst [hunk 2]
```diff
@@ -120,12 +120,27 @@ Maps, Rules and Adapters
       class, the `converters` parameter to the constructor has to be used
       instead.
 
+   .. attribute:: _matcher
```
Comment (davidism) [inline docs/routing.rst:123]: Let's leave this as an internal, non-public API. I don't want to say we're going to support both matchers going forward, as I don't know what future improvements will require.

**Comment by davidism** [comment]
It would be really nice if `BaseConverter.part_isolating` was `True` by default. No one using custom converters now will know to update to it, and will likely be unaware of the change, so they'll always be forced into the less performant code path.

**Comment by pgjones** [comment]
I've changed the default, and added a note in the docs

File: tests/test_routing.py (hunk 1 above)
Comment (pgjones) [reply tests/test_routing.py]: I've re-added this, I now understand it and it is correct if map has redirect_defaults set.

**Pull request merged**